File: java/base/src/main/vsdk/toolkit/io/FileLocator.java
package vsdk.toolkit.io;

import java.io.File;
import java.util.ArrayList;

public class FileLocator {
    private ArrayList<String> searchPaths;

    public FileLocator() {
        searchPaths = new ArrayList<String>();
    }

    public void clearSearchPaths() {
        searchPaths.clear();
    }

    public void addSearchPath(String path) {
        searchPaths.add(path);
    }

    public ArrayList<String> getSearchPaths() {
        return searchPaths;
    }

    public File locate(String filename) {
        File candidate = new File(filename);
        if (candidate.canRead()) {
            return candidate;
        }

        for (int i = 0; i < searchPaths.size(); i++) {
            candidate = new File(searchPaths.get(i) + File.separator + filename);
            if (candidate.canRead()) {
                return candidate;
            }
        }
        return null;
    }
}

File: cpp/base/src/main/vsdk/toolkit/io/FileLocator.cpp
#include <cstdio>

#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/io/FileLocator.h"

void
FileLocator::clearSearchPaths()
{
    searchPaths.clear();
}

void
FileLocator::addSearchPath(const char *path)
{
    searchPaths.add(java::String(path));
}

const java::ArrayList<java::String> &
FileLocator::getSearchPaths() const
{
    return searchPaths;
}

java::File *
FileLocator::locate(const char *filename) const
{
    java::File *candidate = new java::File(filename);
    if (candidate->canRead()) {
        return candidate;
    }
    delete candidate;

    for (long int i = 0; i < searchPaths.size(); i++) {
        char pathname[512];
        snprintf(pathname, sizeof(pathname), "%s/%s", searchPaths.get(i).toCString(), filename);
        candidate = new java::File(pathname);
        if (candidate->canRead()) {
            return candidate;
        }
        delete candidate;
    }
    return nullptr;
}
